File: js/src/record/deserializeRecord.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.deserializeRecord = void 0;
const buffer_1 = require("buffer");
const bs58_1 = require("bs58");
const web3_js_1 = require("@solana/web3.js");
const base_1 = require("@scure/base");
const punycode_1 = require("punycode");
const ipaddr_js_1 = require("ipaddr.js");
const record_1 = require("../types/record");
const error_1 = require("../error");
const checkSolRecord_1 = require("./checkSolRecord");
const trimNullPaddingIdx = (buffer) => {
    const arr = Array.from(buffer);
    const lastNonNull = arr.length - 1 - arr.reverse().findIndex((byte) => byte !== 0);
    return lastNonNull + 1;
};
/**
 * This function can be used to deserialize the content of a record (V1). If the content is invalid it will throw an error
 * @param registry The name registry state object of the record being deserialized
 * @param record The record enum being deserialized
 * @param recordKey The public key of the record being deserialized
 * @returns
 */
const deserializeRecord = (registry, record, recordKey) => {
    const buffer = registry === null || registry === void 0 ? void 0 : registry.data;
    if (!buffer)
        return undefined;
    if (buffer.compare(buffer_1.Buffer.alloc(buffer.length)) === 0)
        return undefined;
    const size = record_1.RECORD_V1_SIZE.get(record);
    const idx = trimNullPaddingIdx(buffer);
    if (!size) {
        const str = buffer.slice(0, idx).toString("utf-8");
        if (record === record_1.Record.CNAME || record === record_1.Record.TXT) {
            return (0, punycode_1.decode)(str);
        }
        return str;
    }
    // Handle SOL record first whether it's over allocated or not
    if (record === record_1.Record.SOL) {
        const encoder = new TextEncoder();
        const expectedBuffer = buffer_1.Buffer.concat([
            buffer.slice(0, 32),
            recordKey.toBuffer(),
        ]);
        const expected = encoder.encode(expectedBuffer.toString("hex"));
        const valid = (0, checkSolRecord_1.checkSolRecord)(expected, buffer.slice(32, 96), registry.owner);
        if (valid) {
            return (0, bs58_1.encode)(buffer.slice(0, 32));
        }
    }
    // Old record UTF-8 encoded
    if (size && idx !== size) {
        const address = buffer.slice(0, idx).toString("utf-8");
        if (record === record_1.Record.Injective) {
            const decoded = base_1.bech32.decodeToBytes(address);
            if (decoded.prefix === "inj" && decoded.bytes.length === 20) {
                return address;
            }
        }
        else if (record === record_1.Record.BSC || record === record_1.Record.ETH) {
            const prefix = address.slice(0, 2);
            const hex = address.slice(2);
            if (prefix === "0x" && buffer_1.Buffer.from(hex, "hex").length === 20) {
                return address;
            }
        }
        else if (record === record_1.Record.A || record === record_1.Record.AAAA) {
            if ((0, ipaddr_js_1.isValid)(address)) {
                return address;
            }
        }
        throw new error_1.InvalidRecordDataError("The record data is malformed");
    }
    if (record === record_1.Record.ETH || record === record_1.Record.BSC) {
        return "0x" + buffer.slice(0, size).toString("hex");
    }
    else if (record === record_1.Record.Injective) {
        return base_1.bech32.encode("inj", base_1.bech32.toWords(buffer.slice(0, size)));
    }
    else if (record === record_1.Record.A || record === record_1.Record.AAAA) {
        return (0, ipaddr_js_1.fromByteArray)([...buffer.slice(0, size)]).toString();
    }
    else if (record === record_1.Record.Background) {
        return new web3_js_1.PublicKey(buffer.slice(0, size)).toString();
    }
    throw new error_1.InvalidRecordDataError("The record data is malformed");
};
exports.deserializeRecord = deserializeRecord;

File: python/src/record/deserialize_record.py
import base64
import ipaddress
import base58
from bech32 import bech32_decode, convertbits, bech32_encode
from solders.pubkey import Pubkey

from NameRegistryState import NameRegistryState
from exception import InvalidRecordDataException
from record.check_sol_record import check_sol_record
from custom_types.record import Record, RECORD_V1_SIZE


def _trim_null_padding_idx(buffer: bytes) -> int:
    """Find the last non-null byte in the buffer."""
    return len(buffer) - next(
        (i for i, byte in enumerate(reversed(buffer)) if byte != 0), len(buffer)
    )


def _is_valid_ip(address: str) -> bool:
    """Validate if the provided string is a valid IP address."""
    try:
        ipaddress.ip_address(address)
        return True
    except ValueError:
        return False


def _process_utf8_record(buffer: bytes, idx: int, record: Record) -> str:
    """Process UTF-8 encoded records like CNAME and TXT."""
    str_data = buffer[:idx].decode("utf-8")
    if record in [Record.CNAME, Record.TXT]:
        return base64.b64decode(str_data).decode("utf-8")
    return str_data


def _process_sol_record(
    buffer: bytes, record_key: Pubkey, registry: NameRegistryState
) -> str | None:
    """Handle the special case for SOL records."""
    expected_buffer = buffer[:32] + bytes(record_key)
    expected = expected_buffer.hex().encode("utf-8")
    valid = check_sol_record(expected, buffer[32:96], registry.owner)
    if valid:
        return base58.b58encode(buffer[:32]).decode("utf-8")
    return None


def _process_old_record(buffer: bytes, idx: int, record: Record) -> str | None:
    """Process old records with specific formats."""
    address = buffer[:idx].decode("utf-8")
    if record == Record.Injective:
        hrp, data = bech32_decode(address)
        if not data:
            raise InvalidRecordDataException("The record data is malformed")
        res = convertbits(data, 5, 8, False)
        if hrp == "inj" and res and len(bytes(res)) == 20:
            return address
    elif record in [Record.BSC, Record.ETH]:
        if address.startswith("0x") and len(bytes.fromhex(address[2:])) == 20:
            return address
    elif record in [Record.A, Record.AAAA]:
        if _is_valid_ip(address):
            return address
    raise InvalidRecordDataException("The record data is malformed")


def _process_current_record(buffer: bytes, size: int, record: Record) -> str | None:
    """Process modern record formats."""
    if record in [Record.ETH, Record.BSC]:
        return "0x" + buffer[:size].hex()
    elif record == Record.Injective:
        bech32_data = convertbits(buffer[:size], 8, 5, True)
        if not bech32_data:
            raise InvalidRecordDataException("The record data is malformed")
        return bech32_encode("inj", bech32_data)
    elif record in [Record.A, Record.AAAA]:
        return str(ipaddress.ip_address(buffer[:size]))
    elif record == Record.Background:
        return str(Pubkey(buffer[:size]))
    raise InvalidRecordDataException("The record data is malformed")


def deserialize_record(
    registry: NameRegistryState | None,
    record: Record,
    record_key: Pubkey,
) -> str | None:
    """
    Deserialize a record based on its type and format.
    """
    if not registry:
        return None
    buffer = registry.data
    if not buffer or buffer == bytes(len(buffer)):
        return None

    size = RECORD_V1_SIZE.get(record)
    idx = _trim_null_padding_idx(buffer)

    # Handle UTF-8 encoded records
    if not size:
        return _process_utf8_record(buffer, idx, record)

    # Handle SOL record (special case)
    if record == Record.SOL:
        sol_result = _process_sol_record(buffer, record_key, registry)
        if sol_result:
            return sol_result

    # Handle old record formats
    if size and idx != size:
        return _process_old_record(buffer, idx, record)

    # Handle modern record formats
    return _process_current_record(buffer, size, record)
